x-pack/plugins/ml/public/application/components/scatterplot_matrix/scatterplot_matrix.tsx

```diff
@@ -0,0 +1,313 @@
+
+import { ml } from '../../services/ml_api_service';
+
+import { getProcessedFields } from '../data_grid';
+
+import {
+  getScatterplotMatrixVegaLiteSpec,
+  LegendType,
+  OUTLIER_SCORE_FIELD,
+} from './scatterplot_matrix_vega_lite_spec';
+
+import './scatterplot_matrix.scss';
+
+const SCATTERPLOT_MATRIX_DEFAULT_FIELDS = 4;
+const SCATTERPLOT_MATRIX_DEFAULT_FETCH_SIZE = 1000;
+const SCATTERPLOT_MATRIX_DEFAULT_FETCH_MIN_SIZE = 1;
+const SCATTERPLOT_MATRIX_DEFAULT_FETCH_MAX_SIZE = 10000;
+
+const TOGGLE_ON = i18n.translate('xpack.ml.splom.toggleOn', {
+  defaultMessage: 'On',
+});
+const TOGGLE_OFF = i18n.translate('xpack.ml.splom.toggleOff', {
+  defaultMessage: 'Off',
+});
+
+const sampleSizeOptions = [100, 1000, 10000].map((d) => ({ value: d, text: '' + d }));
+
+interface ScatterplotMatrixProps {
+  fields: string[];
+  index: string;
+  resultsField?: string;
+  color?: string;
+  legendType?: LegendType;
+}
+
+export const ScatterplotMatrix: FC<ScatterplotMatrixProps> = ({
+  fields: allFields,
+  index,
+  resultsField,
+  color,
+  legendType,
+}) => {
+  const defaultFields =
+    allFields.length > SCATTERPLOT_MATRIX_DEFAULT_FIELDS
+      ? allFields.slice(0, SCATTERPLOT_MATRIX_DEFAULT_FIELDS)
+      : allFields;
+
+  // dynamicSize is optionally used for outlier charts where the scatterplot marks
+  // are sized according to outlier_score
+  const [dynamicSize, setDynamicSize] = useState<boolean>(false);
+
+  // used to give the use the option to customize the fields used for the matrix axes
+  const [fields, setFields] = useState<string[]>(defaultFields);
+
+  // the amount of documents to be fetched
+  const [fetchSize, setFetchSize] = useState<number>(SCATTERPLOT_MATRIX_DEFAULT_FETCH_SIZE);
+  // flag to add a random score to the ES query to fetch documents
+  const [randomizeQuery, setRandomizeQuery] = useState<boolean>(false);
+
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+
+  // contains the fetched documents and columns to be passed on to the Vega spec.
+  const [splom, setSplom] = useState<{ items: any[]; columns: string[] } | undefined>();
+
+  // formats the array of field names for EuiComboBox
+  const fieldOptions = useMemo(
+    () =>
+      allFields.map((d) => ({
+        label: d,
+      })),
+    [allFields]
+  );
+
+  const fieldsOnChange = (newFields: EuiComboBoxOptionOption[]) => {
+    setFields(newFields.map((d) => d.label));
+  };
+
+  const fetchSizeOnChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
+    setFetchSize(
+      Math.min(
+        Math.max(parseInt(e.target.value, 10), SCATTERPLOT_MATRIX_DEFAULT_FETCH_MIN_SIZE),
+        SCATTERPLOT_MATRIX_DEFAULT_FETCH_MAX_SIZE
+      )
+    );
+  };
+
+  const randomizeQueryOnChange = () => {
+    setRandomizeQuery(!randomizeQuery);
+  };
+
+  const dynamicSizeOnChange = () => {
+    setDynamicSize(!dynamicSize);
+  };
+
+  useEffect(() => {
+    async function fetchSplom(options: { didCancel: boolean }) {
```

Comment:
would be better to initialize a callback outside of `useEffect`

Reply:
Can you explain what you'd see as the benefit? This way `fetchSplom` is private inside `useEffect` and can only used within it.

Reply:
Yes, and as I understand this function will re-initialize each time the `useEffect` callback is invoked

Reply:
Yes, but I don't think that's a problem, outside it would reinitialized on every render instead (another option would be to completely move it out of the component, but then it would require a factory or passing on all variables it accesses from the scope). Found this from the React team: `The canonical way to do it is to define an async function inside and call it.` https://github.com/facebook/react/issues/15814#issuecomment-501267882